drivers/serial/uart_shell.c

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright (c) 2024 Yishai Jaffe <yishai1999@gmail.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/drivers/uart.h>
+#include <zephyr/shell/shell.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(uart_shell, CONFIG_LOG_DEFAULT_LEVEL);
+
+static int cmd_uart_baudrate(const struct shell *sh, size_t argc, char **argv)
+{
+	char *s_dev_name = argv[1];
+	const struct device *dev;
+	struct uart_config cfg;
+	uint32_t baudrate;
+	int ret;
+
+	dev = device_get_binding(s_dev_name);
+	if (!dev) {
+		shell_error(sh, "UART: Device driver %s not found.", s_dev_name);
+		return -ENODEV;
+	}
+
+	baudrate = strtol(argv[2], NULL, 10);
+	ret = uart_config_get(dev, &cfg);
+	if (ret < 0) {
+		shell_error(sh, "UART: Failed to get current configuration: %d", ret);
+		return ret;
+	}
+	cfg.baudrate = baudrate;
+
+	ret = uart_configure(dev, &cfg);
+	if (ret < 0) {
+		shell_error(sh, "UART: Failed to configure device: %d", ret);
+		return ret;
+	}
+	return 0;
+}
+
+static int cmd_uart_flow_control(const struct shell *sh, size_t argc, char **argv)
+{
+	char *s_dev_name = argv[1];
+	const struct device *dev;
+	struct uart_config cfg;
+	uint8_t flow_control;
+	int ret;
+
+	dev = device_get_binding(s_dev_name);
+	if (!dev) {
+		shell_error(sh, "UART: Device driver %s not found.", s_dev_name);
+		return -ENODEV;
+	}
+
+	if (!strcmp(argv[2], "none")) {
+		flow_control = UART_CFG_FLOW_CTRL_NONE;
+	} else if (!strcmp(argv[2], "rtscts")) {
+		flow_control = UART_CFG_FLOW_CTRL_RTS_CTS;
+	} else if (!strcmp(argv[2], "dtrdsr")) {
+		flow_control = UART_CFG_FLOW_CTRL_DTR_DSR;
+	} else if (!strcmp(argv[2], "rs485")) {
+		flow_control = UART_CFG_FLOW_CTRL_RS485;
+	} else {
+		shell_error(sh, "Unknown: '%s'", argv[2]);
+		shell_help(sh);
+		return SHELL_CMD_HELP_PRINTED;
+	}
+
+	ret = uart_config_get(dev, &cfg);
+	if (ret < 0) {
+		shell_error(sh, "UART: Failed to get current configuration: %d", ret);
+		return ret;
+	}
+	cfg.flow_ctrl = flow_control;
+
+	ret = uart_configure(dev, &cfg);
+	if (ret < 0) {
+		shell_error(sh, "UART: Failed to configure device: %d", ret);
+		return ret;
+	}
+	return 0;
+}
+
+static void device_name_get(size_t idx, struct shell_static_entry *entry)
+{
+	const struct device *dev = shell_device_lookup(idx, NULL);
```

Comment:
We can use the device wrappers to filter only uart devices.

```suggestion
static bool device_is_uart_and_ready(const struct device *dev)
{
	return DEVICE_API_IS(uart, dev) && device_is_ready(dev);
}

static void device_name_get(size_t idx, struct shell_static_entry *entry)
{
	const struct device *dev = shell_device_filter(idx, device_is_uart_and_ready);
```

Reply:
What do you think about adding a general shell function/macro called `shell_device_lookup_by_api` that will do this for us without needing to define an extra function in each `x_shell.c` module? BTW, accroding to the `shell_device_filter` docs:
```Devices that failed to initialize or do not have a non-empty name are excluded from the candidates for a match.```
So, I think the `device_is_ready` check here is redundant.

Reply:
You need the expanded macro `DEVICE_API_IS` for each type in place, so I don't see how you would be able to generalise.

If `device_is_ready` is already fulfilled, you can drop the check here.

Reply:
We should be able to create a macro so that we can do the following:
```const struct device *dev = SHELL_DEVICE_FILTER_API(idx, uart);```
Don't you think?

Reply:
I'd suggest creating a separate PR to demonstrate.

Reply:
For sure. I'm not certain that it would be as simple as I might have made it sound. I was more asking your opinion. If it is possible it would be pretty convenient to have.

Reply:
I can't really think of how this macro would look like, and I don't think you can make it much simpler as the current implementation.
But if you have an idea, go ahead.